File: app/src/main/java/eu/kanade/tachiyomi/util/RxPager.java
package eu.kanade.tachiyomi.util;

import android.util.Pair;

import java.util.List;

import rx.Observable;
import rx.functions.Func1;
import rx.subjects.PublishSubject;

public class RxPager<T> {

    private final PublishSubject<List<T>> results = PublishSubject.create();
    private int requestedCount;

    public Observable<Pair<Integer, List<T>>> results() {
        requestedCount = 0;
        return results.map(list -> Pair.create(requestedCount++, list));
    }

    public Observable<List<T>> request(Func1<Integer, Observable<List<T>>> networkObservable) {
        return networkObservable.call(requestedCount).doOnNext(results::onNext);
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/util/RxPager.kt
package eu.kanade.tachiyomi.util

import android.util.Pair
import rx.Observable
import rx.subjects.PublishSubject

class RxPager<T> {

    private val results = PublishSubject.create<List<T>>()
    private var requestedCount: Int = 0

    fun results(): Observable<Pair<Int, List<T>>> {
        requestedCount = 0
        return results.map { Pair(requestedCount++, it) }
    }

    fun request(networkObservable: (Int) -> Observable<List<T>>) =
        networkObservable(requestedCount).doOnNext { results.onNext(it) }

}

